packages/adblocker-electron/adblocker.ts

```diff
@@ -7,17 +27,125 @@ export function fromElectronDetails({
-  url: string;
-  resourceType: ElectronRequestType;
-  referrer: string;
-}): Request {
+}: Electron.OnBeforeRequestDetails | Electron.OnHeadersReceivedDetails): Request {
   return Request.fromRawDetails({
     sourceUrl: referrer,
-    type: resourceType || 'other',
+    type: (resourceType || 'other') as ElectronRequestType,
     url,
   });
 }
 
+/**
+ * Wrap `FiltersEngine` into a Electron-friendly helper class.
+ */
+export class ElectronBlocker extends FiltersEngine {
+  public enableBlockingInSession(ses: Electron.Session) {
+    ses.webRequest.onHeadersReceived({ urls: ['<all_urls>'] }, this.onHeadersReceived);
+    ses.webRequest.onBeforeRequest({ urls: ['<all_urls>'] }, this.onBeforeRequest);
+    ipcMain.on('get-cosmetic-filters', this.onGetCosmeticFilters);
+    ses.setPreloads([join(__dirname, './content.js')]);
+  }
+
+  private onGetCosmeticFilters = (
+    e: Electron.IpcMessageEvent,
+    url: string,
+    id: string,
+    msg: IBackgroundCallback & { action?: string },
+  ): void => {
+    // Extract hostname from sender's URL
+    const parsed = parse(url);
+    const hostname = parsed.hostname || '';
+    const domain = parsed.domain || '';
+
+    const { active, styles, scripts } = this.getCosmeticsFilters({
+      domain,
+      hostname,
+      url,
+
+      classes: msg.classes,
+      hrefs: msg.hrefs,
+      ids: msg.ids,
+
+      // This needs to be done only once per frame
+      getBaseRules: msg.lifecycle === 'start',
+      getInjectionRules: msg.lifecycle === 'start',
+      getRulesFromHostname: msg.lifecycle === 'start',
+
+      // This will be done every time we get information about DOM mutation
+      getRulesFromDOM: msg.lifecycle === 'dom-update',
+    });
+
+    if (active === false) {
+      return;
+    }
+
+    // Inject custom stylesheets
+    this.injectStyles(e.sender, styles);
+
+    // Inject scripts from content script
+    e.sender.send(`get-cosmetic-filters-response-${id}`, {
+      active,
+      extended: [],
+      scripts,
+      styles: '',
+    } as IMessageFromBackground);
+  }
+
+  private onHeadersReceived = (
+    details: Electron.OnHeadersReceivedDetails,
+    callback: (a: Electron.Response) => void,
+  ): void => {
+    const policies: string[] = [];
+    const responseHeaders: Electron.ResponseHeaders = details.responseHeaders || {};
+
+    if (details.resourceType === 'mainFrame' || details.resourceType === 'subFrame') {
+      const CSP_HEADER_NAME = 'content-security-policy';
+      const rawCSP = this.getCSPDirectives(fromElectronDetails(details));
+      if (rawCSP !== undefined) {
+        policies.push(...(rawCSP === undefined ? [] : rawCSP.split(';').map(csp => csp.trim())));
+
+        // Collect existing CSP headers from response
+        for (const [name, value] of Object.entries(responseHeaders)) {
+          if (name.toLowerCase() === CSP_HEADER_NAME) {
+            policies.push(value);
+            // @ts-ignore
+            responseHeaders[name] = undefined;
```

Comment:
It would be nice to remove the `// @ts-ignore` directives from this function. I got confused about some Electron types the other day but probably all of it can be done in a type-safe way.